### PR TITLE
LibGUI: Rank emoji search results based on query scores

### PR DESCRIPTION
**Pull request by ebanner**
This PR aims to improve emoji search in the emoji input dialog.

When I type `fire` I'd ideally like for the fire emoji 🔥 to be the first one that comes up.

# Before

image

As you can see the fire emoji 🔥 is way down on the list. That is because the current logic doesn't do any re-ranking based on the query - it just filters away emojis which don't match the query (keeps the original ordering).

# After

image

Now the fire emoji 🔥 ranks highly up on the list because we re-rank by the fuzzy scores.

## Notes

~~Unfortunately fire 🔥 is not the first emoji that comes up. This is because of the scores that `fuzzy_match()` is returning:~~

```
Fire Engine, 🚒 Score: 168
Flag: Ireland, 🇮🇪 Score: 166
Fire, 🔥 Score: 160
```

~~I basically consider this a bug in AK/FuzzyMatch.cpp and I think it's worth opening an issue for it. I expect that fire should get the highest score.~~

EDIT: Looks like this issue has been fixed!

### CODE REVIEW

File: Userland/Libraries/LibGUI/EmojiInputDialog.h [hunk 1]
```diff
@@ -28,9 +28,15 @@ class EmojiInputDialog final : public Dialog {
 private:
     explicit EmojiInputDialog(Window* parent_window);
 
+    struct EmojiScore {
+        Emoji* emoji;
+        int score;
+    };
```
[the remark this answers is not in the document]
Comment (ebanner) [reply Userland/Libraries/LibGUI/EmojiInputDialog.h:34]: > This struct is only used in `filter_and_rank_emojis`, no need to define it here

Hm, if we define `EmojiScore` in EmojiInputDialog.cpp, then we won't be able to implement the suggestion https://github.com/SerenityOS/serenity/pull/24644#discussion_r1699150840 as we'll need to specify the signature as `Vector<EmojiScore> EmojiInputDialog::filter_and_rank_emojis(ByteString const& query)` in the header file, which we won't be able to do, right?

**Comment by LucasChollet** [comment]
Seems alright, we don't squash commits on merge tho, please fixup your second commit into the first one.

**Comment by ebanner** [comment]
done 👍

File: Userland/Libraries/LibGUI/EmojiInputDialog.h [hunk 2]
```diff
@@ -28,9 +28,15 @@ class EmojiInputDialog final : public Dialog {
 private:
     explicit EmojiInputDialog(Window* parent_window);
 
+    struct EmojiScore {
+        Emoji* emoji;
```
Comment (kleinesfilmroellchen) [inline Userland/Libraries/LibGUI/EmojiInputDialog.h:32]: Can be a reference.

**Review by kleinesfilmroellchen — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibGUI/EmojiInputDialog.h (hunk 2 above)
Comment (LucasChollet) [reply Userland/Libraries/LibGUI/EmojiInputDialog.h:32]: I already pointed that out earlier.
https://discord.com/channels/830522505605283862/830739873119207426/1268611093232750703

**Pull request closed without merging**